### PR TITLE
Test3d

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
A bit early, but I'm fixing up the materials science 3d example. It seems to run, but I haven't ran the program to completion yet.

For the backend side of things, looking at cdgmm3d, I compared against the 2d version, and brought over some of the functionality. cdgmm3d is now inplace. Added back the _fft_convolve method.

For the core/scattering3d side of things, I've been trying to make it do the same things it did before you made the core folder, and  ~~it seems like things are failing due to~~ the finalize method (which is on the back end). It seems like the code I have should be doing the same thing the previous code was doing, ~~but things are failing due to torch.stack in finalize.~~

(THIS error is now fixed)
error it gives me on test files:
```
s_order_1 = [torch.cat([arr[...,0] for arr in scattering_coef], 0) for scattering_coef in s_order_1]
E   RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 0. Got 32 and 16 in dimension 1 at /pytorch/aten/src/TH/generic/THTensor.cpp:689
```
~~Tomorrow in the morning I am going to look in more depth at the old code and compare outputs from each function, and make sure they are doing the same thing.~~

The test file for scattering3d seems to work. There are a bunch of is_complex and complex_modulus  test functions, want me to move them to another test file, delete them, or just hook those methods up to the back end named tuple? ~~The error messages that come from these tests should give more clarity to where my problem is.~~

### CODE REVIEW

**Comment by edouardoyallon** [comment]
Make sure to fix #6 before continuing this PR - then rebase this one on top of it, or I can do it iyw

File: kymatio/scattering3d/backend/torch_backend.py [hunk 1]
```diff
@@ -120,14 +136,34 @@ def modulus_rotation(x, module):
...
+        filter_array: torch tensor
+            size (M, N, O, 2)
+
+        Returns
+        -------
+
+        output: the result of the convolution of input_array with filter
+
+        """
+        return fft(cdgmm3d(fft(input_array, inverse=False), filter_array), inverse=True)
```
Comment (edouardoyallon) [inline kymatio/scattering3d/backend/torch_backend.py]: I'm pros removing fftconvolve, and explicitely perform those operations in `scattering3d.py` (as I started to)

File: kymatio/scattering3d/backend/torch_backend.py [hunk 2]
```diff
@@ -274,14 +310,17 @@ def compute_integrals(input_array, integral_powers):
 
 
 
-backend = namedtuple('backend', ['name', 'cdgmm3d', 'fft', 'Pad', 'finalize',
-    'modulus', 'modulus_rotation', 'subsample', 'compute_integrals'])
+backend = namedtuple('backend', ['name', 'cdgmm3d', 'fft', 'finalize',
+     'to_complex', 'modulus', 'modulus_rotation', 'averaging', 'subsample', 'compute_integrals'])
```
Comment (edouardoyallon) [inline kymatio/scattering3d/backend/torch_backend.py]: cf #6

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 3]
```diff
@@ -97,12 +99,62 @@ def build(self):
...
+                self.backend._compute_local_scattering_coefs(x,
+                        self.guassian_filters[j+1], self.points)
         elif self.method == 'standard':
-            self.average = lambda x, j: compute_global (x, self.points, j)
+            self.average = lambda x, j:\
+                self.backend._compute_standard_scattering_coefs(x,
+                        self.gaussian_filters[j], self.J, self.backend.subsample)
+
+    
+    def _apply(self, fn):
```
Comment (edouardoyallon) [inline kymatio/scattering3d/frontend/torch_frontend.py]: those functions will be removed(they will be inherited via `torch`)

File: examples/3d/scattering3d_qm7.py [hunk 4]
```diff
@@ -11,7 +11,8 @@
```
Comment (edouardoyallon) [inline examples/3d/scattering3d_qm7.py]: for now, as the examples are not part of the test suit, feel free to modify later this file, I think it's better to focus on the frontend-backend

**Comment by edouardoyallon** [comment]
I think the issue is right now next to the function `averaging`... something is not consistent with the original code, but I'm confident we'll be able to track it!

File: kymatio/scattering3d/core/scattering3d.py [hunk 5]
```diff
@@ -44,16 +44,16 @@ def scattering3d(_input, filters, gaussian_filters, rotation_covariant, points,
...
             if max_order >1:
                 for j_2 in range(j_1+1, J+1):
-                    U_2_c = fft(U_1_m)
+                    frequency_2 = fft(U_1_m)
                     U_2_m = None
 
                     if rotation_covariant:
                         for m in range(len(filters[l][j_1])):
-                            U_2_c = cdgmm3d(U_2_c, filters[l][j_1][m])
+                            U_2_c = cdgmm3d(frequency_2, filters[l][j_1][m])
```
Comment (edouardoyallon) [inline kymatio/scattering3d/core/scattering3d.py]: is there a reason to call it `frequency_2`(any name is arbitrary) - would be great to think of it later, @janden maybe has some ideas

**Comment by edouardoyallon** [comment]
thanks for catching the bug!

**Comment by edouardoyallon** [comment]
I realized that, actually, the FFT of several signals was computed J^2 too much time. I think it could be good to think of feeding "averaging" with the fft of the signal instead of the raw signal

**Comment by MuawizChaudhary** [comment]
Assuming that we are feeding in the FFT of the signal to the averaging function, does this make our compute_integrals method incorrect now? Should I do fft(input, inverse=True) at the start of that method then? or is it fine?

I fixed up the methods that used in selecting the average, with the assumption in mind that we are feeding in the FFT of the signal. I believe it should be correct now, I followed [this](https://reader.elsevier.com/reader/sd/pii/S2213158217300384?token=EF254A06958DC127CD66B4AD4FE437D1E2E8B918B3CFD572EB6ADD25C18E790D3227CB657904E58A8CD1F572763143AB) paper when working on the averaging functions and incorporating the Gaussian filters. It would be nice if you could make sure that I did things correctly!

I also worked on adding the filters to buffers. I know you said you would take care of it yourself, but I am bad at listening to certain instructions :) (Do tell me if I did something wrong with regards to the buffer, but it does seem to work). Only 4 tests fail, and these are the same tests that fail in the current master branch of Kymatio for the 3d scattering. 

How do we want to proceed now?

**Comment by edouardoyallon** [comment]
> Assuming that we are feeding in the FFT of the signal to the averaging function, does this make our compute_integrals method incorrect now? Should I do fft(input, inverse=True) at the start of that method then? or is it fine?

I think feeding the signal in the spatial domain is better than the frequency domain, and trying to make the call to fft visible in core() is, I think, a good thing. We might refactor later

> I fixed up the methods that used in selecting the average, with the assumption in mind that we are feeding in the FFT of the signal. I believe it should be correct now, I followed [this](https://reader.elsevier.com/reader/sd/pii/S2213158217300384?token=EF254A06958DC127CD66B4AD4FE437D1E2E8B918B3CFD572EB6ADD25C18E790D3227CB657904E58A8CD1F572763143AB) paper when working on the averaging functions and incorporating the Gaussian filters. It would be nice if you could make sure that I did things correctly!

```
=================================================================================================== test session starts ===================================================================================================
platform darwin -- Python 3.5.6, pytest-3.8.1, py-1.6.0, pluggy-0.11.0
rootdir: /Users/oyallon/kymatio, inifile:
plugins: cov-2.7.1
collected 11 items                                                                                                                                                                                                        

tests/test_torch_scattering3d.py ......FF..                                                                                                                                                                         [ 90%]
tests/test_utils_scattering3d.py .  
```

Looks OK but the tests still do not pass

> I also worked on adding the filters to buffers. I know you said you would take care of it yourself, but I am bad at listening to certain instructions :P (Do tell me if I did something wrong with regards to the buffer, but it does seem to work). Only 4 tests fail, and these are the same tests that fail in the current master branch of Kymatio for the 3d scattering.
> 
> How do we want to proceed now?

Fixing all the tests will be the key

**Comment by MuawizChaudhary** [comment]
Alright, so I got all the unit tests to pass on my end, please make sure this is the case on your end.

One thing that should be solved is the torch dependency in core, I needed to add this back in order for things to work. Your decision on if we want to create appropriate methods in the backend for our purposes. 

I compute an inverse FFT when feeding into the compute integrals function in order to get the correct result. let me know if there is a more efficient way of doing this. 

One thing to note, we can't do scattering.integral_powers = integral_powers or the like anymore, the integral_powers must be defined and an argument to the scattering object now, something you might want to keep in mind when you refactor.

**Pull request merged**